Data type browser

ApurvaSinha17 opened the pull request:
Added code for Data type browser

ApurvaSinha17 committed:
Data type Browser code

ApurvaSinha17 committed:
Data type Browser code

zednis commented:
Let's complete pull request #2 and then merge the new master into this branch so the diffs in this pull request are only related to the datatype browser.

edit - I'll talk with @mrpatrickwest to make sure we close out #2 soon.

ApurvaSinha17 committed:
Made some corrections in datatypes.xml

ApurvaSinha17 committed:
Made some corrections in datatypes.xml

ApurvaSinha17 committed:
Merge remote-tracking branch 'origin/data-type-browser' into data-type-browser

zednis commented on opensearch/descriptions/datatypes.xml:
```diff
@@ -0,0 +1,30 @@
+  document describes how to call the various RESTful services in order
+  to fill s2s facets (select creation year of the data type, creator of the data type, parameters of the data type, source standard of the data type, subject area of the data type, and the result set.)</Description>
+  <Tags></Tags>
+  <Contact>Apurva Kumar Sinha</Contact>
```
change contact to web@mail.deepcarbon.net

ApurvaSinha17 replied:
changed

zednis commented on opensearch/descriptions/datatypes.xml:
```diff
@@ -0,0 +1,30 @@
+  <Url type="text/html" rel="dco:ResultsListQuery" template="http://data.deepcarbon.net/dco-s2s/opensearch/services/datatypes/search.php?request=datatypes&amp;creator={dco:Creator?}&amp;parameter={dco:Parameter?}&amp;sourcestandard={dco:SourceStandard?}&amp;subjectarea={dco:SubjectArea?}&amp;creationyear={dco:CreatedAtTime?}&amp;limit={count}&amp;offset={startIndex}" />
+
+  <LongName>DCO Data Portal Datatype Search Service</LongName>
+  <Developer>Apurva Kumar Sinha</Developer>
```
Change developer to "DCO Data Science Team"

ApurvaSinha17 replied:
changed

zednis commented:
let's use  camelCase on parameter names in the open search service.

"subjectarea"  -> "subjectArea"
"sourcestandard" -> "sourceStandard"
 "creationyear" -> "creationYear"

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+		switch($type) {
+			case "datatypes":
+				$header .= "?datatype ?dco_id ?label ?year ?parameterLabel ?access ";
+				$header .= '(GROUP_CONCAT(DISTINCT ?creator ; SEPARATOR=",") AS ?creator) ';
```
I am not sure we should be defining a new variable (i.e. ?creator, ?creator_label, ?sourceStandard, ?sourceStandard_label)  with the same name as the existing variable.

See [project.php](https://github.com/tetherless-world/dco-s2s-browsers/blob/master/opensearch/services/projects/projects.php) as an example.

ApurvaSinha17 committed:
Modified parameter names to have camelCase and changed value in contact and developer elements

ApurvaSinha17 committed:
Modified parameter names to have camelCase

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+				$body .= "BIND(str(?t) AS ?year) . ";
+				break;
+				
+			case "creator":
```
I am pretty sure this query is broken.  Have you tested it?
1. the creator instance should not have type dco:creator
2. the creator instance should not have the role instance as a type
3. this query should include the ?id and ?label variables (see other comments below)

ApurvaSinha17 replied:
Made corrections

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+				$body .= "BIND(str(?l) AS ?sourceStandard_label) . ";
+				break;
+
+			case "subjectArea":
```
The expected parameters to return for this query should be 

`?id ?label (COUNT(DISTINCT ?datatype) AS ?count)`

and yet the query body does not contain ?id or ?label.

Rewrite to

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype vivo:hasSubjectArea ?id . ";
$body .= "?id skos:prefLabel ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

ApurvaSinha17 replied:
Should it be like this? I checked the SPARQL query using this and the results were coming.

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype dco:dataTypeSubjectArea ?id . ";
$body .= "?id a skos:Concept . ";
$body .= "?id rdfs:label ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

zednis replied:
If we proceed with using dco:dataTypeSubjectArea then that would be the correct query.

edit - are we using skos:prefLabel with the subject area concepts or just rdfs:label?

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+				$body .= "BIND(str(?c) AS ?creator) . ";
+				break;
+
+			case "parameter":
```
The expected parameters to return for this query should be 

`?id ?label (COUNT(DISTINCT ?datatype) AS ?count)`

and yet the query body does not contain ?id or ?label.

Rewrite to

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype dco:hasParameter ?id . ";
$body .= "?id rdfs:label ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+		
+		$body = "";
+		switch($type) {
+			case "creationYear":
```
1. You should add a case statement to the header for ?year (see comment above)

ApurvaSinha17 replied:
added

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+			case "count":
+				$header .= "(count(DISTINCT ?datatype) AS ?count)";
+				break;
+			default:
```
Add a new case statement for when the query parameter is "creationYear".  In this case we are not return ?id and ?label but just ?year.  You will likewise need to add a case for "creationYear" in the getQueryFooter function.

```
case "creationYear":
    $header .= '?year' ;
    break;
```

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+				break;
+			case "count":
+				break;
+			default:
```
Add a case for "creationYear".

```
case "creationYear":
    $footer .= " GROUP BY ?year";
    break;
```

zednis commented on opensearch/services/datatypes/datatypes.php:
```diff
@@ -0,0 +1,430 @@
+				$body .= "BIND(str(?mt) AS ?modifiedTime) . ";
+				$body .= "BIND(str(?sl) AS ?sourceStandard_label) . ";
+				$body .= "BIND(str(?use) AS ?use) . ";
+				$body .= "BIND(str(?pl) AS ?parameterLabel) . ";
```
Can there be more than 1 parameter per datatype?

ApurvaSinha17 replied:
Yes, there is one sample data type instance in UDCO that has multiple parameters.

ApurvaSinha17 committed:
Corrections in datatypes.php as per the review comments

ApurvaSinha17 committed:
Made datatypes.php consistent with the version in UDCO

Commit:
Modified the queries for the result set and the year. Modified the
constraint for the year to use Filter. Modified the result set output to
use the right returned results.

Commit:
Up to date with new directory structure. Removed context from creation
year.

Commit:
Making the URLs https

Commit:
Merge branch 'master' into data-type-browser

Conflicts:
	ontologies/dco-s2s.ttl

Commit:
http changed to https

Commit:
http changed to https

Commit:
Merge branch 'data-type-browser' of https://github.com/tetherless-world/dco-s2s-browsers into data-type-browser

mrpatrickwest commented:
Merged master into this branch, updated to use https, pushed to udco (changed urls there) and it works. Ready to merge the pull request.

Commit:
Merge remote-tracking branch 'origin/master' into data-type-browser

Commit:
Made the init script executable on checkout. Updated the endpoint for
upper case VIVO
